### PR TITLE
Add rate counter

### PR DESCRIPTION
**Pull request by stephenbez**
Note: clock is from https://github.com/facebookgo/clock

### CODE REVIEW

File: rate_counter.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+}
+
+func (this *StandardRateCounter) Mark(n int64) {
+	atomic.AddInt64(&this.counter, n)
+	this.maybeSampleCount()
+}
+
+func (this *StandardRateCounter) Rate1() float64 {
+	this.maybeSampleCount()
+	return this.lastRate
```
Comment (ghost) [inline rate_counter.go]: race: reading of lastRate is not protected by lock.

File: rate_counter.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+ * May sample the current count and timestamp.  Note that this is not an unbiased sampling
+ * algorithm, but given that we are computing a rate over a ring buffer of 60 samples, it
+ * should not matter in practice.
+ */
+func (this *StandardRateCounter) maybeSampleCount() {
+	var currentTimeMs int64 = this.clock.Now().UnixNano() / 1e6
+	currentSampleTimeMs := this.roundTime(currentTimeMs)
+
+	this.lock.RLock()
+	if currentSampleTimeMs > this.lastSampleTimeMs {
```
Comment (ghost) [inline rate_counter.go]: nit: can you refactor the code a bit to,
```
this.lock.RLock()
toSample := currentSampleTimeMs > this.lastSampleTimeMs
this.lock.RUnlock()
if !toSample {
  return
}

this.lock.Lock()
defer this.lock.Unlock()
...
```

**Review by ghost — APPROVED** [review]

File: rate_counter.go (hunk 2 above)
Comment (stephenbez) [reply rate_counter.go]: Done.

File: rate_counter.go (hunk 1 above)
Comment (stephenbez) [reply rate_counter.go]: I now make maybeSampleCount (which holds the lock) return lastRate.

**Review by ghost — APPROVED** [review]

**Pull request merged**